pkg/challenges/challenges.go

```diff
@@ -291,3 +355,34 @@ func isSpiffeIDAllowed(host, spiffeID string) bool {
 	}
 	return strings.Contains(u.Hostname(), "."+host)
 }
+
+// isURISubjectAllowed compares the subject and issuer URIs,
+// returning an error if the scheme or the hostnames do not match
+func isURISubjectAllowed(subject, issuer *url.URL) error {
+	subjectHostname := subject.Hostname()
+	issuerHostname := issuer.Hostname()
+
+	if subject.Scheme != issuer.Scheme {
+		return fmt.Errorf("Subject (%s) and Issuer (%s) URI schemes do not match", subject.Scheme, issuer.Scheme)
+	}
+
+	// If the hostnames exactly match, return early
+	if subjectHostname == issuerHostname {
+		return nil
+	}
+
+	// Compare the top level and second level domains
+	sHostname := strings.Split(subjectHostname, ".")
+	iHostname := strings.Split(issuerHostname, ".")
+	if len(sHostname) < 2 {
+		return fmt.Errorf("Subject URI hostname too short: %s", subjectHostname)
+	}
+	if len(iHostname) < 2 {
+		return fmt.Errorf("Issuer URI hostname too short: %s", issuerHostname)
+	}
+	if sHostname[len(sHostname)-1] == iHostname[len(iHostname)-1] &&
+		sHostname[len(sHostname)-2] == iHostname[len(iHostname)-2] {
+		return nil
+	}
+	return fmt.Errorf("Subject and issuer hostnames do not match: %s, %s", subjectHostname, issuerHostname)
```

Comment:
Nit: `Issuer` is capitalised in other error messages.

Reply:
Fixed